l3kernel/l3intarray.dtx

```diff
@@ -173,7 +173,454 @@
+%   to count the elements first. Otherwise it is basically the same as
+%   \cs{intarray_new:Nn} (which we don't use because it would complain about the
+%   wrong qualifier in debugging mode)
+%   This is not implemented in Lua to ensure that the clist parsing is consistent
+%   with the clist module.
+%    \begin{macrocode}
+\cs_new_protected:Npn \intarray_const_from_clist:Nn #1#2
+  {
+    \@@_new:N #1
+    \@@_gset_count:Nw #1 \clist_count:n {#2} \scan_stop:
+    \int_zero:N \l_@@_loop_int
+    \clist_map_inline:nn {#2}
+      {
+        \int_incr:N \l_@@_loop_int
+        \__kernel_intarray_gset:Nnn #1 \l_@@_loop_int { \int_eval:n {##1} } }
+  }
+\cs_generate_variant:Nn \intarray_const_from_clist:Nn { c }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}[rEXP]{\intarray_to_clist:N, \intarray_to_clist:c}
+% \begin{macro}[rEXP]{\@@_to_clist:Nn, \@@_to_clist:w}
+%   Loop through the array, putting a comma before each item.  Remove
+%   the leading comma with |f|-expansion.  We also use the auxiliary in
+%   \cs{intarray_show:N} with argument comma, space.
+%    \begin{macrocode}
+\cs_new:Npn \intarray_to_clist:N #1 { \@@_to_clist:Nn #1 { , } }
+\cs_generate_variant:Nn \intarray_to_clist:N { c }
+%</tex>
+%<*lua>
+local concat = table.concat
+luacmd('@@_to_clist:Nn', function()
+  local t = @@_table()
+  local sep = token.scan_string()
+  sprint(-2, concat(t, sep))
+end, 'global')
+%</lua>
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% \begin{macro}[rEXP]{\__kernel_intarray_range_to_clist:Nnn, \@@_range_to_clist:ww}
+%   Loop through part of the array.
+%    \begin{macrocode}
+%<*tex>
+\cs_new:Npn \__kernel_intarray_range_to_clist:Nnn #1#2#3
+  {
+    \exp_after:wN \@@_range_to_clist:ww
+    \int_value:w \int_eval:w #2 \exp_after:wN ;
+    \int_value:w \int_eval:w #3
+    #1
+  }
+%</tex>
+%<*lua>
+luacmd('@@_range_to_clist:ww', function()
+  local from = scan_int()
+  scan_token()
+  local to = scan_int()
+  local t = @@_table()
+  sprint(-2, concat(t, ',', from, to))
+end, 'global')
+%</lua>
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\__kernel_intarray_gset_range_from_clist:Nnn, \@@_gset_range:nNw}
+%   Loop through part of the array. We allow additional commas at the end.
+%    \begin{macrocode}
+%<*tex>
+\cs_new_protected:Npn \__kernel_intarray_gset_range_from_clist:Nnn #1#2#3
+  {
+    \@@_gset_range:w \int_eval:w #2 #1 #3 , , \scan_stop:
+  }
+%</tex>
+%<*lua>
+luacmd('@@_gset_range:w', function()
+  local from = scan_int()
+  local t = @@_table()
+  while true do
+    local tok = scan_token()
+    if tok == comma then
+      repeat
+        tok = scan_token()
+      until tok ~= comma
+      break
+    else
+      put_next(tok)
+    end
+    t[from] = scan_int()
+    scan_token()
+    from = from + 1
+  end
+  end, 'global', 'protected')
+%</lua>
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\intarray_show:N, \intarray_show:c, \intarray_log:N, \intarray_log:c}
```

Comment:
The definition of `\intarray_show:N` etc seems identical in all engines, should be shared.

Reply:
The definition of `\@@_show:NN` will probably be slightly different once the second space at the beginning is fixed, therefore it has been kept separate. The `\intarray_show:N` etc. could still be merged but then `\@@_show:NN` would need separate documentation which would effectively just be a textual version of the definition of `\intarray_show:N` and `\intarray_log:N`, so it seemed useful to keep them here too.

Reply:
I think the fix should be in `\@@_to_clist:Nn` for non-LuaTeX, such as adding an argument `#4` to `\@@_to_clist:w`, which is initially `\use_none:n` (when called from the main `_to_clist:Nn` and later is `\exp_not:n` (when called from `_to_clist:w`).  Many other ways obviously.

Reply:
Sounds reasonable, so I moved the `show` functions to the shared part.